File: src/main/java/com/gcu/MessageService/data/DataAccessInterface.java
package com.gcu.MessageService.data;

import java.util.List;
import java.util.Optional;

public interface DataAccessInterface<T>{

    public List<T> getAll();

    public List<T> findAll(String id);

    /**
     * A method to create an object in the database
     * @param t The object to be persisted to the database
     * @return int The number of affected database rows
     */
    public Optional<T> create(T t);

    public Optional<T> update(T t);

    /**
     * Removes an object from the database
     * @param t The object to be removed
     * @return int The number of affected database rows
     */
    public boolean delete(T t);
}

File: src/main/kotlin/com/gcu/messageservice/data/DataAccessInterface.kt
package com.gcu.messageservice.data

interface DataAccessInterface<T> {

    fun getAll(): List<T>
    fun findAllWithID(id: String): List<T>
    fun create(t: T): T?
    fun update(t: T): T?
    fun delete(t: T): Boolean
}